// @nodegui/nodegui/dist/lib/QtCore/QSizeF.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QSizeF = void 0;
const Component_1 = require("../core/Component");
const addon_1 = __importDefault(require("../utils/addon"));
const helpers_1 = require("../utils/helpers");
class QSizeF extends Component_1.Component {
    constructor(nativeOrWidth, height) {
        let native;
        if (!nativeOrWidth) {
            native = new addon_1.default.QSizeF();
        }
        else if (helpers_1.checkIfNativeElement(nativeOrWidth)) {
            native = nativeOrWidth;
        }
        else {
            native = new addon_1.default.QSizeF(nativeOrWidth, height);
        }
        super(native);
    }
    setWidth(width) {
        return this.native.setWidth(width);
    }
    width() {
        return this.native.width();
    }
    setHeight(height) {
        return this.native.setHeight(height);
    }
    height() {
        return this.native.height();
    }
    static fromQVariant(variant) {
        return new QSizeF(addon_1.default.QSizeF.fromQVariant(variant.native));
    }
}
exports.QSizeF = QSizeF;
